src/transformers/parallelization_utils.py

```diff
@@ -0,0 +1,1098 @@
+        pass
+
+    @staticmethod
+    def attn_qkv(layer, config):
+        return []
+
+    @staticmethod
+    def attn_out(layer, config):
+        return []
+
+    @staticmethod
+    def mlp_in(layer, config):
+        return []
+
+    @staticmethod
+    def mlp_out(layer, config):
+        return []
+
+    @staticmethod
+    def word_embedding(model, config):
+        return []
+
+    @staticmethod
+    def layerwise_copy_to_all(layer, config):
+        return []
+
+    @staticmethod
+    def modelwise_copy_to_all(model, config):
+        return []
+
+    @staticmethod
+    def original_layer_class():
+        raise NotImplementedError
+
+
+class ParallelEngine(object):
+    """
+    Integration with Parallelformers
+    https://github.com/tunib-ai/parallelformers
+    """
+
+    def __init__(
+        self,
+        policy: ParallelismPolicy,
+        hold_params: bool,
+        skip_bias_add: bool,
+        vocab_parallel_embedding: bool,
+        additional_layers: List[ParallelLayer] = None,
+    ):
+        self.hold_params = hold_params
+        self.skip_bias_add = skip_bias_add
+        self.vocab_parallel_embedding = vocab_parallel_embedding
+        self.device = torch.cuda.current_device()
+        self.policy = policy
+        self.param_dict = {}
+
+        if additional_layers is None:
+            self.additional_layers = []
+        else:
+            self.additional_layers = additional_layers
+
+    def parallelize(self, model, mpu):
+        config = model.config
+
+        if dist.is_initialized() and mpu is not None:
+            gpu_index = mpu.get_tensor_model_parallel_rank()
+            world_size = mpu.get_tensor_model_parallel_world_size()
+        else:
+            gpu_index = 0
+            world_size = 1
+
+        self._process_additional_layers(
+            model=model,
+            mpu=mpu,
+            additional_layers=self.additional_layers,
+            config=config,
+        )
+
+        self._process_repeated_layers(
+            model=model,
+            mpu=mpu,
+            gpu_index=gpu_index,
+            world_size=world_size,
+            config=config,
+        )
+
+        if self.vocab_parallel_embedding:
+            self._process_word_embedding(
+                model=model,
+                mpu=mpu,
+                gpu_index=gpu_index,
+                world_size=world_size,
+                config=config,
+            )
+
+        non_cuda_tensors = []
+        for k, v in dict(model.state_dict()).items():
+            if not v.is_cuda:
+                if torch.is_tensor(v):
+                    non_cuda_tensors.append(k)
+
+        if len(non_cuda_tensors) > 0:
+            raise Exception(f"{non_cuda_tensors} are not CUDA tensors now.")
+
+    def _process_word_embedding(self, model, mpu, gpu_index, world_size, config):
+        hidden_size = config.hidden_size
+        original_vocab_size = config.vocab_size
+
+        if hasattr(config, "make_vocab_size_divisible_by"):
```

Comment:
It's for the megatron-friendly models (e.g. `ParallelGPT2`)
check this: https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/arguments.py#L296